[touchwand] Touchwand Binding initial contribution

roieg opened the pull request:
<!--
Thanks for contributing to the openHAB project!
Please describe the goal and effect of your PR here.
Pay attention to the below notes and to *the guidelines* for this repository.
Feel free to delete any comment sections in the template (starting with "<!--").
-->

<!-- TITLE -->
TouchWand Wanderfull controller binding initial contribution
<!--
Please provide a PR summary in the *Title* above, according to the following schema:
- If related to one specific add-on: Mention the add-on shortname in square brackets
  e.g. "[exec]", "[netatmo]" or "[tesla]"
- If the PR is work in progress: Add "[WIP]"
- Give a short meaningful description in imperative mood
  e.g. "Add support for device XYZ" or "Fix wrongly handled exception"
  for a new add-on/binding: "Initial contribution"
Examples:
- "[homematic] Improve communication with weak signal devices"
- "[timemachine][WIP] Initial contribution"
- "Update contribution guidelines on new signing rules"
-->
[touchwand][WIP] initial contribution 
Touchwand Wanderfull™ Hub basic is a plug & play Z-Wave based controller that uses Wi-Fi and Bluetooth to easily connect all smart home components.
The binding allow to connect to Touchwand Wanderfull hub and control shutters and switches.
[TouchWand website](http://www.touchwand.com/products/touchwand-wanderfull-hub/)
This is initial contribution , currently the binding supports
* Hub manual adding
* Units auto discovery 
supported units type -
1. Shutters
2. Switches



 
<!-- DESCRIPTION -->

<!--
Please give a few sentences describing the overall goals of the pull request.
Give enough details to make the improvement and changes of the PR understandable
to both developers and tech-savy users.

Please keep the following in mind:
- What is the classification of the PR, e.g. Bugfix, Improvement, Novel Addition, ... ?
- Did you describe the PRs motivation and goal?
- Did you provide a link to any prior discussion, e.g. an issue or community forum thread?
- Did you describe new features for the end user?
- Did you describe any noteworthy changes in usage for the end user?
- Was the documentation updated accordingly, e.g. the add-on README?
- Does your contribution follow the coding guidelines:
  https://www.openhab.org/docs/developer/development/guidelines.html
- Did you check for any (relevant) issues from the static code analysis:
  https://www.openhab.org/docs/developer/development/bindings.html#static-code-analysis
- Did you sign-off your work:
  https://www.openhab.org/docs/developer/contributing/contributing.html#sign-your-work
-->

<!-- TESTING -->

<!--
Your Pull Request will automatically be built and available under the following folder:
https://openhab.jfrog.io/openhab/libs-pullrequest-local/org/openhab/

It is a good practice to add a URL to your built JAR in this Pull Request description,
so it is easier for the community to test your Add-on.
If your Pull Request contains a new binding, it will likely take some time
before it is reviewed and processed by maintainers.
That said, consider submitting your Add-on in the Eclipse IoT Marketplace
See this thread for more info:
https://community.openhab.org/t/24491

Don't forget to submit a thread about your Add-on in the openHAB community:
https://community.openhab.org/c/add-ons 
-->

Hilbrand commented:
Git usage went a bit wrong here. To recover:
```
git reset --hard b26eda3fc8dd5588824ce2f2855d8c9ea6f8ebb5
git pull --rebase origin master  <-- assuming origin points to `openhab/openhab2-addons
```
Then cherrypick all commits that went in between (These should be all commits):
```
git cherry-pick 219062b3e3a54e3104627bd25d8cae210e2bea41
git cherry-pick 6db4cc0ebc18e89b646872ca6e7538b0f51ac598
git cherry-pick 84ba70262762cc798505cd7ef704e1f12e70641e
git cherry-pick 4b142c69594dac2d3c38d9666ad3685892640fd3
git cherry-pick 33627cba43a578f0aef5c1d611f527e15a289632
```
Then push back
```
git push --force-with-lease
```

roieg commented:
> I think you've got an issue with this PR that touches sooooo many files. You should rebase.

Apology , I forgot to signoff my commits and when I followed the instructions to fix this it attached all other commits since my PR.
I dont know how to fix it

roieg commented:
> Git usage went a bit wrong here. To recover:
> 
> ```
> git reset --hard b26eda3fc8dd5588824ce2f2855d8c9ea6f8ebb5
> git pull --rebase origin master  <-- assuming origin points to `openhab/openhab2-addons
> ```
> 
> Then cherrypick all commits that went in between (These should be all commits):
> 
> ```
> git cherry-pick 219062b3e3a54e3104627bd25d8cae210e2bea41
> git cherry-pick 6db4cc0ebc18e89b646872ca6e7538b0f51ac598
> git cherry-pick 84ba70262762cc798505cd7ef704e1f12e70641e
> git cherry-pick 4b142c69594dac2d3c38d9666ad3685892640fd3
> git cherry-pick 33627cba43a578f0aef5c1d611f527e15a289632
> ```
> 
> Then push back
> 
> ```
> git push --force-with-lease
> ```

Thanks , will try

Hilbrand commented:
In `ESH-INF` xml files all must be tabs. In Eclipse do `Ctrl-Shift-F` in the file to fix this. Alternative a pom.xml should have spaces.

roieg commented:
hi @Hilbrand  @clinique 
is there any addition active action needed from my side to this pull request ?

Hilbrand commented:
@roieg There will be a more detailed review. You can already take a look at the report generated by maven. See the travis-ci build or run maven locally on your binding. Most of these messages should be fixed. Fixing and pay attention to details in your binding code will make the review process go faster as we don't have to mention it anymore.

To format your code you can apply the spotless plugin. This was recently added, see: https://www.openhab.org/docs/developer/guidelines.html#b-code-formatting-rules-style

roieg commented:
@Hilbrand  , any idea why PR-openHAB-Addons build fails ?
It complains on " ERROR: Couldn't find any revision to build. Verify the repository and branch configuration for this job"

Hilbrand commented:
I've changed the base branch to merge this pr against to 2.5.x. This could explain it. Also you need to update/rebase the pom version to 2.5.2-SNAPSHOT. Don't forget to add your binding to the CODEOWNERS file.

roieg commented:
> I've changed the base branch to merge this pr against to 2.5.x. This could explain it. Also you need to update/rebase the pom version to 2.5.2-SNAPSHOT. Don't forget to add your binding to the CODEOWNERS file.

Hi @Hilbrand  ,I updated the pom file and retried but the build still fails
Any idea what am i doing wrong ?

fwolter commented:
Are you still interested in a review/merge? I would take a look into it, then.

roieg commented:
@fwolter  thank you for replying 
I am still interested , please take a look

roiegeron-spearuav commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandBaseUnitHandler.java:
```diff
@@ -0,0 +1,177 @@
+
+        if (!bridgeStatus.equals(ThingStatus.ONLINE)) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE);
+            logger.debug("UpdateStatus OFFLINE BRIDGE_OFFLINE {}", getThing().getLabel());
```
Done

roiegeron-spearuav commented on bundles/org.openhab.binding.touchwand/README.md:
```diff
@@ -0,0 +1,39 @@
+
+## Supported Things
+
+TouhWand Hub supports switches , shutters and scenarios configured in touchwand hub controller
```
fixed

roiegeron-spearuav commented on bundles/org.openhab.binding.touchwand/README.md:
```diff
@@ -0,0 +1,39 @@
+
+TouhWand Hub supports switches , shutters and scenarios configured in touchwand hub controller
+
+## control 
```
fixed

roiegeron-spearuav commented on bundles/org.openhab.binding.touchwand/README.md:
```diff
@@ -0,0 +1,39 @@
+
+## Discovery
+
+After adding TouchWand Hnb the auto discovery will add all switches , shutters and scenarios to the inbox.
```
fixed

roieg replied:
done

roiegeron-spearuav commented on bundles/org.openhab.binding.touchwand/README.md:
```diff
@@ -0,0 +1,39 @@
+
+## Binding Configuration
+
+**Touchwand Wanderfull™** Hub Controller need to be added manually by IP address. The controller require **username** and **password**  
```
fixed

roiegeron-spearuav commented on bundles/org.openhab.binding.touchwand/README.md:
```diff
@@ -0,0 +1,39 @@
+
+No thing configuration is needed
+
+## Full Example
+
+TBD
```
added

roiegeron-spearuav commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandAlarmSensorHandler.java:
```diff
@@ -0,0 +1,54 @@
+
+    public TouchWandAlarmSensorHandler(Thing thing) {
+        super(thing);
+        Instant.now().toEpochMilli();
```
I removed it from the code as Alarm Sensor is not currently fully supported

roiegeron-spearuav commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandAlarmSensorHandler.java:
```diff
@@ -0,0 +1,54 @@
+    @Override
+    void updateTouchWandUnitState(TouchWandUnitData unitData) {
+        Integer batteryLevel = ((TouchWandUnitDataAlarmSensor) unitData).getCurrStatus().getBatt();
+        DecimalType dBatteryLevel = DecimalType.valueOf(String.valueOf(batteryLevel));
```
removed from the code

roiegeron-spearuav commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandBaseUnitHandler.java:
```diff
@@ -0,0 +1,177 @@
+
+    @Override
+    public void dispose() {
+        logger.trace("Handler disposed.");
```
replaced

roiegeron-spearuav replied:
removed

roiegeron-spearuav commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandBaseUnitHandler.java:
```diff
@@ -0,0 +1,177 @@
+    public void initialize() {
+        ThingStatus bridgeStatus;
+        logger.trace("Initializing TocuhWand Unit handler");
+        updateStatus(ThingStatus.UNKNOWN);
```
moved just before job scheduling

roiegeron-spearuav commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandBaseUnitHandler.java:
```diff
@@ -0,0 +1,177 @@
+        logger.trace("initializeThing Thing {} Bridge status {}", getThing().getUID(), bridgeStatus);
+
+        Thing thing = getThing();
+        unitId = thing.getUID().toString().split(":")[3]; // touchwand id
```
combined to one line

roiegeron-spearuav commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandBindingConstants.java:
```diff
@@ -0,0 +1,81 @@
+    public static final String TYPE_DIMMER = "dimmer";
+    public static final String TYPE_ALARMSENSOR = "AlarmSensor";
+
+    public static final String[] SUPPORTED_TOCUHWAND_TYPES = { TYPE_WALLCONTROLLER, TYPE_SWITCH, TYPE_SHUTTER,
```
could not find the typo....

roiegeron-spearuav commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandBindingConstants.java:
```diff
@@ -0,0 +1,81 @@
+        SUPPORTED_THING_TYPES_UIDS.add(THING_TYPE_SWITCH);
+        SUPPORTED_THING_TYPES_UIDS.add(THING_TYPE_SHUTTER);
+        SUPPORTED_THING_TYPES_UIDS.add(THING_TYPE_WALLCONTROLLER);
+        SUPPORTED_THING_TYPES_UIDS.add(THING_TYPE_DIMMER); // not implemented yet
+        SUPPORTED_THING_TYPES_UIDS.add(THING_TYPE_ALARMSENSOR); // not implemented yet
```
dimmer is implemented , I removed the comment
commented ALARMSENSOR

roiegeron-spearuav commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandBridgeHandler.java:
```diff
@@ -0,0 +1,177 @@
+
+    @Override
+    public void initialize() {
+        updateStatus(ThingStatus.UNKNOWN);
```
done

roiegeron-spearuav commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandBridgeHandler.java:
```diff
@@ -0,0 +1,177 @@
+
+        try {
+            @SuppressWarnings("unused")
+            InetAddress addr = InetAddress.getByName(config.get(HOST).toString()); // validate IP address
```
ok , I removed the IP address validation

roieg commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandBaseUnitHandler.java:
```diff
@@ -0,0 +1,177 @@
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        if (command instanceof RefreshType) {
+            // updateTouchWandUnitState(getUnitState(unitId));
```
I still consider if it is the right place to update the unitStatus , I left the code commented till further decision ...

roieg commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandDimmerHandler.java:
```diff
@@ -0,0 +1,64 @@
+        switch (command.toString()) {
+            case "OFF":
+            case "ON":
+                bridgeHandler.touchWandClient.cmdSwitchOnOff(unitId, (OnOffType) command);
```
in Dimmer the value can be brightness as well
for simplicity I used the same channel , I can later separate it to ON/OFF and brightness channels

fwolter commented:
I received notifications for some responses by you, but I don't see them in the comments. Did you remove some messages or is Github messing something up?

roieg commented:
> I received notifications for some responses by you, but I don't see them in the comments. Did you remove some messages or is Github messing something up?

Hi @fwolter I have been doing some work today and responded specifically on many on your detailed reviews point (closed , done and so)

fwolter commented:
Ok, then Github seemed to mess something up. I'll respond to your comments from my notifications.

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandBaseUnitHandler.java:
```diff
@@ -0,0 +1,169 @@
+
+        bridgeStatus = bridge.getStatus();
+
+        unitId = getThing().getUID().toString().split(":")[3]; // touchwand id
```
Actually I meant this:
```suggestion
        unitId = getThing().getUID().getId();
```

roieg replied:
yep , much more elegant .
I changed it and commited

roieg commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandRestClient.java:
```diff
@@ -0,0 +1,195 @@
+    private final boolean cmdLogin(String user, String pass, String ipAddr) {
+        String command = buildUrl(CMD_LOGIN) + "user=" + user + "&" + "psw=" + pass;
+        String response = sendCommand(command, METHOD_GET, null);
+        if (response != null && !response.equals("Unauthorized")) {
+            return true;
+        }
+        return false;
```
changed

roieg commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandWallControllerHandler.java:
```diff
@@ -0,0 +1,59 @@
+public class TouchWandWallControllerHandler extends TouchWandBaseUnitHandler {
+
+    private long timeSinceLastEvent;
+    private final int ajdustentEventFilterTime = 2000; // 2 seconds
```
changed ajdustentEventFilterTime to static

roieg commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandControllerDiscoveryService.java:
```diff
@@ -0,0 +1,148 @@
+    }
+
+    private void addDeviceDiscoveryResult(String label, String ip) {
+        String id = ip.replaceAll("\\.", "");
```
Wouldn't 
192.168.0.1 -> 19216801 
192.168.1.1 -> 19216811

roieg commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/data/AlarmSensorCurrStatus.java:
```diff
@@ -0,0 +1,99 @@
+    // private Double sensor1;
+    // @SerializedName("alrm_snsr")
+    // @Expose
+    // private AlrmSnsr alrmSnsr;
```
Currently I removed support for Alarm sensors

fwolter commented:
Do you raise your hand, when you incorporated the review feedback?

roieg commented:
@fwolter  I am not sure I understand the question

fwolter commented:
When you finished making all changes, please send a message. I will review your changes, then.

roieg commented:
@fwolter Sure , I still have some changes I would like to incorporate. 
I will let you know

roieg commented on bundles/org.openhab.binding.touchwand/src/main/resources/ESH-INF/thing/alarmsensor.xml:
```diff
@@ -0,0 +1,25 @@
+		</channels>		
+	</thing-type>
+	
+	<channel-type id="sensor_alarm">
```
I removed alarmsensor.xml from the code as it is not supported at the moment

roieg commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandBridgeHandler.java:
```diff
@@ -0,0 +1,177 @@
+        try {
+            @SuppressWarnings("unused")
+            InetAddress addr = InetAddress.getByName(config.get(HOST).toString()); // validate IP address
+            host = config.get(HOST).toString();
```
@fwolter   what would be the best behavior in case the mandatory configuration parameters are null?

fwolter replied:
The Thing should be set to OFFLINE with status detail CONFIGURATION_ERROR and a message that explains what's wrong.

roieg commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandRestClient.java:
```diff
@@ -0,0 +1,195 @@
+    private final boolean cmdLogin(String user, String pass, String ipAddr) {
+        String command = buildUrl(CMD_LOGIN) + "user=" + user + "&" + "psw=" + pass;
+        String response = sendCommand(command, METHOD_GET, null);
+        if (response != null && !response.equals("Unauthorized")) {
```
I cleaned the code , return value is bool based on the response

roieg commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandRestClient.java:
```diff
@@ -0,0 +1,195 @@
+        try {
+            url = new URL(command);
+        } catch (MalformedURLException e) {
+            logger.warn("Error building URL {} : {}", command, e.getMessage());
```
@fwolter which exception should I throw ? the same one ?

fwolter replied:
Yes, it's ok to throw the same. You can remove the try/catch.

fwolter replied:
Can you comment on this?

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandBridgeHandler.java:
```diff
@@ -0,0 +1,159 @@
+                .registerService(DiscoveryService.class.getName(), discoveryService, new Hashtable<String, Object>()));
+    }
+
+    @SuppressWarnings("null")
```
Can this suppression be removed?

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandDimmerHandler.java:
```diff
@@ -0,0 +1,64 @@
+        switch (command.toString()) {
+            case "OFF":
+            case "ON":
+                bridgeHandler.touchWandClient.cmdSwitchOnOff(unitId, (OnOffType) command);
```
It's good to use the same Channel. What I meant is to check if it is really an OnOffType. Because the user can also send StringType with "ON"/"OFF" into this Channel, which would result in a ClassCastException. This prevents it:
```suggestion
                if (command instanceof OnOffType) {
                    bridgeHandler.touchWandClient.cmdSwitchOnOff(unitId, (OnOffType) command);
                }
```

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandHandlerFactory.java:
```diff
@@ -0,0 +1,86 @@
+ */
+
+@Component(configurationPid = "binding.touchwand", service = ThingHandlerFactory.class)
+@NonNullByDefault({})
```
```suggestion
@NonNullByDefault
```

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandRestClient.java:
```diff
@@ -0,0 +1,196 @@
+    @SuppressWarnings("null")
+    private final boolean cmdLogin(String user, String pass, String ipAddr) {
+        String command = buildUrl(CMD_LOGIN) + "user=" + user + "&" + "psw=" + pass;
+        String response = sendCommand(command, METHOD_GET, "");
```
You should annotate `sendCommand()` with `@NonNull` if it can return a null value. Then, the null suppression above could be removed.

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandRestClient.java:
```diff
@@ -0,0 +1,196 @@
+
+    private static final String CONTENT_TYPE_APPLICATION_JSON = MimeTypes.Type.APPLICATION_JSON.asString();
+
+    private static final int REQUEST_TIMEOUT_SEC = 10000; // 10 seconds
```
```suggestion
    private static final int REQUEST_TIMEOUT_SEC = 10;
```

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandRestClient.java:
```diff
@@ -0,0 +1,196 @@
+        }
+
+        request = httpClient.newRequest(url.toString()).timeout(REQUEST_TIMEOUT_SEC, TimeUnit.SECONDS).method(method);
+        if (method.equals(METHOD_POST) && (content != "")) {
```
See above
```suggestion
        if (method.equals(METHOD_POST) && (!content.isEmpty())) {
```

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandShutterHandler.java:
```diff
@@ -0,0 +1,69 @@
+
+    @Override
+    void touhWandUnitHandleCommand(Command command) {
+        switch (command.toString()) {
```
This can be left as is. Please ignore my comment.

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandWallControllerHandler.java:
```diff
@@ -0,0 +1,58 @@
+@NonNullByDefault
+public class TouchWandWallControllerHandler extends TouchWandBaseUnitHandler {
+
+    private long timeSinceLastEvent;
```
It's good practice to append the unit to the field name e.g. timeSinceLastEventMs.

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandWebSockets.java:
```diff
@@ -0,0 +1,203 @@
+ *
+ * @author Roie Geron - Initial contribution
+ */
+@NonNullByDefault({})
```
```suggestion
@NonNullByDefault
```

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandWebSockets.java:
```diff
@@ -0,0 +1,203 @@
+ * @author Roie Geron - Initial contribution
+ */
+@NonNullByDefault({})
+public class TouchWandWebSockets {
+
+    private WebSocketClient client;
+    private String controllerAddress;
+    private TouchWandSocket touchWandSocket;
+    private boolean isShutDown = false;
+
+    private final Logger logger = LoggerFactory.getLogger(TouchWandWebSockets.class);
+    private List<TouchWandUnitStatusUpdateListener> listeners = new ArrayList<>();
+    @Nullable
+    private ScheduledFuture<?> socketReconnect;
+
+    private static final String WS_ENDPOINT_TOUCHWAND = "/async";
+    private @Nullable URI uri;
+    private static final int CONNECT_TIMEOUT = 10000; // 10 seconds
+    private static final int WEBSOCKET_RECONNECT_INTERVAL = CONNECT_TIMEOUT * 2;
+    private static final int WEBSOCKET_IDLE_TIMEOUT = CONNECT_TIMEOUT * 10;
+    ScheduledExecutorService scheduler;
```
Can you move final and static fields to the top of the class?

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandWebSockets.java:
```diff
@@ -0,0 +1,203 @@
+
+    private final Logger logger = LoggerFactory.getLogger(TouchWandWebSockets.class);
+    private List<TouchWandUnitStatusUpdateListener> listeners = new ArrayList<>();
+    @Nullable
+    private ScheduledFuture<?> socketReconnect;
```
```suggestion
    private @Nullable ScheduledFuture<?> socketReconnect;
```

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandWebSockets.java:
```diff
@@ -0,0 +1,203 @@
+
+    private static final String WS_ENDPOINT_TOUCHWAND = "/async";
+    private @Nullable URI uri;
+    private static final int CONNECT_TIMEOUT = 10000; // 10 seconds
+    private static final int WEBSOCKET_RECONNECT_INTERVAL = CONNECT_TIMEOUT * 2;
+    private static final int WEBSOCKET_IDLE_TIMEOUT = CONNECT_TIMEOUT * 10;
```
It's good practice to append the unit to the field name e.g. TIMEOUT_SEC.

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandWebSockets.java:
```diff
@@ -0,0 +1,203 @@
+    }
+
+    @WebSocket(maxIdleTime = WEBSOCKET_IDLE_TIMEOUT)
+    @NonNullByDefault
```
Inner classes don't need to be annotated with NonNull. There is a bug that checkstyle complains about it. You can safely ignore it.

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandWebSockets.java:
```diff
@@ -0,0 +1,203 @@
+    @NonNullByDefault
+    public class TouchWandSocket {
+
+        @SuppressWarnings("unused")
+        private @Nullable Session session;
```
Does this field has any function or can it be removed?

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandWebSockets.java:
```diff
@@ -0,0 +1,203 @@
+            JsonParser jsonParser = new JsonParser();
+            try {
+                JsonObject unitObj = jsonParser.parse(msg).getAsJsonObject();
+                boolean eventUnitChanged = unitObj.get("type").getAsString().equals("UNIT_CHANGED");
```
`get("type")` can return null. You should add a null check. Same for below.

fwolter replied:
Can you comment on this?

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandWebSockets.java:
```diff
@@ -0,0 +1,203 @@
+
+        @SuppressWarnings("null")
+        private void asyncWeb() {
+            if (socketReconnect == null || socketReconnect.isDone()) {
```
The null suppression can be removed if you store `socketReconnect` to a local variable and do the null check on that.

roieg replied:
I am using a global variable and I am trying to prevent a case where there are more then one websocket open .

fwolter replied:
As you are using a global variable, you should store the reference into a local variable and do the nullcheck and the further processing in this method on that, because the global variable could be set to null between `socketReconnect == null` and `socketReconnect.isDone()`.

roieg replied:
Made the change , hopefully correctly

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandControllerDiscoveryService.java:
```diff
@@ -0,0 +1,145 @@
+ * @author Roie Geron - Initial contribution
+ */
+@Component(service = DiscoveryService.class, immediate = false, configurationPid = "discovery.touchwand")
+@NonNullByDefault({})
```
```suggestion
@NonNullByDefault

```

fwolter replied:
Can you comment on this?

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandControllerDiscoveryService.java:
```diff
@@ -0,0 +1,145 @@
+@NonNullByDefault({})
+public class TouchWandControllerDiscoveryService extends AbstractDiscoveryService {
+
+    private static final int SEARCH_TIME = 2;
```
It's good practice to append the unit to the field name e.g. TIMEOUT_SEC.

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandControllerDiscoveryService.java:
```diff
@@ -0,0 +1,145 @@
+    private Thread socketReceiveThread = null;
+    private DatagramSocket listenSocket = null;
+
+    private final Logger logger = LoggerFactory.getLogger(TouchWandControllerDiscoveryService.class);
```
Can you move the final logger to the top of the class?

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandControllerDiscoveryService.java:
```diff
@@ -0,0 +1,145 @@
+
+    @Override
+    public void deactivate() {
+        super.deactivate();
```
Super destructors should be called at the end of the implementing method.

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandControllerDiscoveryService.java:
```diff
@@ -0,0 +1,145 @@
+                .withThingType(THING_TYPE_BRIDGE)
+                .withLabel(label)
+                .withProperties(properties)
+                .withRepresentationProperty(ip)
```
The argument must be the *name* of the property.
```suggestion
                .withRepresentationProperty("ipAddress")

```

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandControllerDiscoveryService.java:
```diff
@@ -0,0 +1,145 @@
+    protected void runReceiveThread(DatagramSocket socket) {
+        socketReceiveThread = new ReceiverThread(socket);
+        socketReceiveThread.start();
+        socketReceiveThread.setName(TouchWandBindingConstants.BINDING_ID);
```
Better set the name before the thread starts.

fwolter replied:
In the meantime the thread naming convention has changed. See #8216 Can you change it to something like this `OH-binding-<bindingId>-discovery`?

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandControllerDiscoveryService.java:
```diff
@@ -0,0 +1,145 @@
+
+        private void receiveData(DatagramPacket datagram) {
+            try {
+                while (true) {
```
You need to check if the thread is interrupted. Otherwise, interrupting in `deactivate()` won't work.
```suggestion
                while (!isInterrupted()) {
```

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandControllerDiscoveryService.java:
```diff
@@ -0,0 +1,145 @@
+                while (true) {
+                    mySocket.receive(datagram);
+                    InetAddress address = datagram.getAddress();
+                    String sentence = new String(dgram.getData(), 0, dgram.getLength(), "US-ASCII");
```
Using a constant will prevent throwing an UnsupportedEncodingException:
```suggestion
                    String sentence = new String(dgram.getData(), 0, dgram.getLength(), StandardCharsets.US_ASCII);
```

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandUnitDiscoveryService.java:
```diff
@@ -0,0 +1,230 @@
+@NonNullByDefault({})
+public class TouchWandUnitDiscoveryService extends AbstractDiscoveryService {
+
+    private static final int SEARCH_TIME = 10;
+    private static final int SCAN_INTERVAL = 60;
+    private static final int LINK_DISCOVERY_SERVICE_INITIAL_DELAY = 5;
```
It's good practice to append the unit to the field name e.g. TIMEOUT_SEC.

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandUnitDiscoveryService.java:
```diff
@@ -0,0 +1,230 @@
+ *
+ * @author Roie Geron - Initial contribution
+ */
+@NonNullByDefault({})
```
```suggestion
@NonNullByDefault
```

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandUnitDiscoveryService.java:
```diff
@@ -0,0 +1,230 @@
+ * @author Roie Geron - Initial contribution
+ */
+@NonNullByDefault({})
+public class TouchWandUnitDiscoveryService extends AbstractDiscoveryService {
```
Can you sort the fields after static, final as above?

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandUnitDiscoveryService.java:
```diff
@@ -0,0 +1,230 @@
+
+        logger.debug("Starting TouchWand discovery on bridge {}", touchWandBridgeHandler.getThing().getUID());
+        String response = touchWandBridgeHandler.touchWandClient.cmdListUnits();
+        if (response == "") {
```
See above
```suggestion
        if (response.isEmpty()) {
```

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandUnitDiscoveryService.java:
```diff
@@ -0,0 +1,230 @@
+
+    @Override
+    protected synchronized void stopScan() {
+        super.stopScan();
+        removeOlderResults(getTimestampOfLastScan());
```
Call super destructor at the end.
```suggestion
        removeOlderResults(getTimestampOfLastScan());
        super.stopScan();
```

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandUnitDiscoveryService.java:
```diff
@@ -0,0 +1,230 @@
+
+    public void activate() {
+        super.activate(null);
+        removeOlderResults(new Date().getTime(), touchWandBridgeHandler.getThing().getUID());
```
This will save some resources. Same for below.
```suggestion
        removeOlderResults(System.currentTimeMillis(), touchWandBridgeHandler.getThing().getUID());
```

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandUnitDiscoveryService.java:
```diff
@@ -0,0 +1,230 @@
+    @Override
+    protected void startBackgroundDiscovery() {
+        if (scanningJob == null || scanningJob.isCancelled()) {
+            scanningJob = scheduler.scheduleWithFixedDelay(scanningRunnable, LINK_DISCOVERY_SERVICE_INITIAL_DELAY,
+                    SCAN_INTERVAL, TimeUnit.SECONDS);
```
Syntactical sugar. Then, the inner class TouchWandUnitScan and the field could be removed.
```suggestion
            scanningJob = scheduler.scheduleWithFixedDelay(this::startScan, LINK_DISCOVERY_SERVICE_INITIAL_DELAY,
                    SCAN_INTERVAL, TimeUnit.SECONDS);
```

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandUnitDiscoveryService.java:
```diff
@@ -0,0 +1,230 @@
+                .withLabel(unit.getName())
+                .withBridge(bridgeUID)
+                .withProperties(properties)
+                .withRepresentationProperty(unit.getId().toString())
```
```suggestion
                .withRepresentationProperty("id")
```

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/dto/Csc.java:
```diff
@@ -0,0 +1,60 @@
+
+    @SerializedName("sceneNo")
+    @Expose
+    private Integer sceneNo;
```
Primitive types should be used where possible.
There are several locations where you use complex types, where `int` would be sufficient. I didn't mark all of them.

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/dto/TouchWandUnitFromJson.java:
```diff
@@ -0,0 +1,74 @@
+    public TouchWandUnitFromJson() {
+    }
+
+    public static TouchWandUnitData parseResponse(JsonObject JsonUnit) {
```
```suggestion
    public static TouchWandUnitData parseResponse(JsonObject jsonUnit) {
```

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/resources/ESH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,10 @@
+	xsi:schemaLocation="https://openhab.org/schemas/binding/v1.0.0 https://openhab.org/schemas/binding-1.0.0.xsd">
+
+	<name>TouchWand Binding</name>
+	<description>Connect to TouchWand Wanderfull™ communicate with units connected to the controller.</description>
```
```suggestion
	<description>Connect to TouchWand Wanderfull™ Hub and communicate with units connected to the controller.</description>
```

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/resources/ESH-INF/config/config.xml:
```diff
@@ -0,0 +1,49 @@
+	<config-description uri="bridge-type:touchwand:bridge">
+		<parameter name="username" type="text">
+			<label>Username</label>
+			<description>Username for TouchWand Wanderfull Hub </description>
```
Same for below
```suggestion
			<description>Username for TouchWand Wanderfull Hub</description>
```

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/resources/ESH-INF/config/config.xml:
```diff
@@ -0,0 +1,49 @@
+			<required>true</required>
+		</parameter>
+		<parameter name="port" type="integer" min="80" max="65535">
+			<context>network-address</context>
```
This must not be a network-address. You can remove it.

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/resources/ESH-INF/config/config.xml:
```diff
@@ -0,0 +1,49 @@
+		<parameter name="addSecondaryUnits" type="boolean">
+			<required>false</required>
+			<default>false</default>
+			<description>If the controller is primary, add secondaries controllers units as well</description>
```
```suggestion
			<description>If the controller is primary, add secondary controllers units as well</description>
```

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/resources/ESH-INF/config/config.xml:
```diff
@@ -0,0 +1,49 @@
+			<required>false</required>
+			<default>false</default>
+			<description>If the controller is primary, add secondaries controllers units as well</description>
+			<label>secondary</label>
```
```suggestion
			<label>Add Secondary Controllers</label>
```

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/resources/ESH-INF/thing/bridge.xml:
```diff
@@ -0,0 +1,14 @@
+	xmlns:thing="https://openhab.org/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">
+	<bridge-type id="bridge">
+		<label>TouchWand Wanderfull Hub bridge</label>
```
```suggestion
		<label>TouchWand Wanderfull Hub Bridge</label>
```

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/resources/ESH-INF/thing/dimmer.xml:
```diff
@@ -0,0 +1,25 @@
+		<label>Dimmer</label>
+		<description>The dimmer channel allows to switch the light on and off and set brightness.
+		</description>
+		<category>Device control (ON / OFF , brightess 0-100%)</category>
```
Did you push your changes? I still see text in the category tag.

roieg commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandBaseUnitHandler.java:
```diff
@@ -0,0 +1,170 @@
+
+        bridgeHandler.registerUpdateListener(this);
+
+        if (!bridgeStatus.equals(ThingStatus.ONLINE)) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE);
+            return;
+        }
```
removed

roieg commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandBaseUnitHandler.java:
```diff
@@ -0,0 +1,170 @@
+    @NonNullByDefault({})
+    protected String unitId;
+    protected final Logger logger = LoggerFactory.getLogger(TouchWandBaseUnitHandler.class);
+    private @Nullable ScheduledFuture<?> pollingJob;
```
removed

roieg commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandBaseUnitHandler.java:
```diff
@@ -0,0 +1,170 @@
+        });
+    }
+
+    private int getUnitState(String unitId) {
```
I keep it for RefreshType , at the moment the unit status response from the hub is not reliable , I will keep the code for future use when they fix the status reliability

roieg commented:
@fwolter  thank you for your hard work and feedback.
I have done most of the changes requested in the review.
There are some changes (very few) changes which I will address in the near future

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandBridgeHandler.java:
```diff
@@ -0,0 +1,159 @@
+            if (thingReachable) {
+                updateStatus(ThingStatus.ONLINE);
+                registerItemDiscoveryService(this);
+                touchWandWebSockets = new TouchWandWebSockets(host, scheduler);
+                touchWandWebSockets.registerListener(this);
+                touchWandWebSockets.connect();
```
You could assign it to a local variable, too, to avoid the compiler warning.
```suggestion
                TouchWandWebSockets localSockets = touchWandWebSockets = new TouchWandWebSockets(host, scheduler);
                localSockets.registerListener(this);
                localSockets.connect();
```

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandControllerDiscoveryService.java:
```diff
@@ -0,0 +1,146 @@
+    protected void runReceiveThread(DatagramSocket socket) {
+        socketReceiveThread = new ReceiverThread(socket);
+        socketReceiveThread.setName(TouchWandBindingConstants.BINDING_ID);
+        socketReceiveThread.start();
```
See above. You can store it to a local variable to eliminate the compiler warning.

roieg replied:
@fwolter 
Done some acrobatics to eliminate all warnings in @NonNullByDefault and all compiler warnings 
in TouchWandControllerDiscoveryService.java
Thread name been changed to match new naming convention 

regarding  unitObj.get("type")  , the "type" is currently always received in the response and  it is very stable . I will look to add some checks , but at this point I prefer to leave it as is if it is ok

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandUnitDiscoveryService.java:
```diff
@@ -0,0 +1,217 @@
+
+    @Override
+    protected void startScan() {
+        if (touchWandBridgeHandler.touchWandClient == null) {
```
This cannot be null, you can remove the null check to eliminate the compiler warning.

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandUnitDiscoveryService.java:
```diff
@@ -0,0 +1,217 @@
+
+    @Override
+    protected void startBackgroundDiscovery() {
+        if (scanningJob == null || scanningJob.isCancelled()) {
```
See above. Store to a local variable to eliminate the compiler warning.

roieg replied:
I been try it and the scanning job failed , for now I will commit the change but doing some more testing to verify it is a valid solution.

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/resources/ESH-INF/config/config.xml:
```diff
@@ -0,0 +1,48 @@
+	xsi:schemaLocation="https://openhab.org/schemas/config-description/v1.0.0 https://openhab.org/schemas/config-description-1.0.0.xsd">
+
+	<config-description uri="bridge-type:touchwand:bridge">
+		<parameter name="username" type="text">
+			<label>Username</label>
+			<description>Username for TouchWand Wanderfull Hub</description>
+			<required>true</required>
```
The required tag is deprecated. Can you use the attribute? Same for below.
```suggestion
		<parameter name="username" type="text" required="true">
			<label>Username</label>
			<description>Username for TouchWand Wanderfull Hub</description>
```

fwolter commented on bundles/org.openhab.binding.touchwand/README.md:
```diff
@@ -0,0 +1,80 @@
+|----------|--|----|--|
+| username|Touchwand hub username  | string | yes
+| password|Touchwand hub password | string | yes
+|ipAddress|Touchwand hub IP address | dot notated IP address | yes
```
```suggestion
|ipAddress|Touchwand hub hostname or IP address | string | yes
```

fwolter commented on bundles/org.openhab.binding.touchwand/README.md:
```diff
@@ -0,0 +1,80 @@
+| password|Touchwand hub password | string | yes
+|ipAddress|Touchwand hub IP address | dot notated IP address | yes
+|port| Management port (default 80) | integer | no
+|statusrefresh| Unit status refresh interval | seconds | no
```
```suggestion
|statusrefresh| Unit status refresh interval in seconds | integer | no
```

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/resources/ESH-INF/thing/dimmer.xml:
```diff
@@ -0,0 +1,24 @@
+		</supported-bridge-type-refs>
+		<label>TouchWand Dimmer Unit</label>
+		<channels>
+			<channel id="brightness" typeId="brightness">
```
You could use the system definition of a dimmer. Then, you don't need the channel-type definition below.
```suggestion
			<channel id="brightness" typeId="system.brightness">
```

fwolter reviewed (changes requested):
There are some formatting issues. You can view them with `mvn spotless:check -Dspotless.check.skip=false` and fix them with `mvn spotless:apply`.

You could rebase your branch to get rid of the checkstyle NonNullByDefault warnings. This is also necessary to specify the 2.5.9-SNAPSHOT version in the pom.xml. This way you could also fix the conflicts.

That's how you do it:
If not already done, add the openHAB addon repo as remote to your repo and fetch it:
```
git remote add upstream https://github.com/openhab/openhab-addons.git
git fetch upstream
```
Then, you can rebase your PR's branch onto `2.5.x`:
```
git rebase upstream/2.5.x
```
Fix the conflicts. Finally force-push the rebased branch to this PR's branch:
```
git push origin touchwand --force-with-lease
```

fwolter commented:
@roieg as the merge window for 2.5.9 comes closer and closer, do you want this get merged before OH3?

roieg commented:
Yes, i will try to finish it this weekend

On Sat, Sep 5, 2020, 12:58 Fabian Wolter <notifications@github.com> wrote:

> @roieg <https://github.com/roieg> as the merge window for 2.5.9 comes
> closer and closer, do you want this get merged before OH3?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openhab/openhab-addons/pull/6310#issuecomment-687582601>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABK6LK5VGSSUWAENHNN2FVTSEIDU3ANCNFSM4JHLX6ZA>
> .
>

roieg commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandBaseUnitHandler.java:
```diff
@@ -0,0 +1,155 @@
+            if (!this.getThing().getStatusInfo().getStatus().equals(ThingStatus.ONLINE)) {
+                updateStatus(ThingStatus.ONLINE);
+            }
+        } catch (JsonParseException | IllegalStateException | UnsupportedOperationException e) {
```
removed

roieg commented:
@fwolter  i been rebaseing to 2.5.9 
wtill need to do mvn spotless:check -Dspotless.check.skip=false 
Having some issues when doing it getting **The platform metadata area could not be writte**n error
Will deal with it ASAP

fwolter commented:
Maybe you must close eclipse before calling maven.

roieg commented:
> ## Travis tests have failed
> Hey @roieg,
> please read the following log in order to understand the failure reason. There might also be some helpful tips along the way.
> It will be awesome if you fix what is wrong and commit the changes.

Hi , 

> Maybe you must close eclipse before calling maven.

Hi , 
Strange , seems that after 2.5.9 merge building a windows environment fails for all bindings  , not just **touchwand** with the same reason
**eclipse wtp formatters - XML' found problem in 'src\main\resources\ESH-INF\binding\binding.xml**

I tried a clean installed environment , using original openhab-addons repository 2.5.x HEAD, and it fails.
I will consult the community .
Anyhow mange to move to linux development environment , run mvn spotless:apply and commit.

I fix most of the issues , the few ones i did not , I commented
Can you please review ?

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandBaseUnitHandler.java:
```diff
@@ -0,0 +1,154 @@
+    public static final Set<ThingTypeUID> SUPPORTED_THING_TYPES = new HashSet<>(
+            Arrays.asList(THING_TYPE_SHUTTER, THING_TYPE_SWITCH, THING_TYPE_WALLCONTROLLER, THING_TYPE_DIMMER));
+
+    @NonNullByDefault({})
+    protected String unitId;
+
+    @NonNullByDefault({})
+    protected TouchWandBridgeHandler bridgeHandler;
```
Please make these `@Nullable`.

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandBaseUnitHandler.java:
```diff
@@ -0,0 +1,154 @@
+                updateStatus(ThingStatus.ONLINE);
+            }
+        } catch (JsonParseException | IllegalStateException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, "Could not parse cmdGetUnitById");
```
You could include the exception's message.
```suggestion
            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, "Could not parse cmdGetUnitById: "+e.getMessage());
```

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandBaseUnitHandler.java:
```diff
@@ -0,0 +1,154 @@
+            }
+        } catch (JsonParseException | IllegalStateException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, "Could not parse cmdGetUnitById");
+            logger.warn("Could not parse cmdGetUnitById response for unit id {}  label {}", unitId,
+                    getThing().getLabel());
```
This log message could be removed, since`updateStatus()` already does the logging.

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandControllerDiscoveryService.java:
```diff
@@ -0,0 +1,141 @@
+        runReceiveThread(localListenSocket);
+    }
+
+    @Override
+    protected synchronized void stopScan() {
+        super.stopScan();
+    }
```
Please remove this

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandControllerDiscoveryService.java:
```diff
@@ -0,0 +1,141 @@
+
+    protected void runReceiveThread(DatagramSocket socket) {
+        Thread localSocketReceivedThread = socketReceiveThread = new ReceiverThread(socket);
+        localSocketReceivedThread.setName(TouchWandBindingConstants.DISCOVERY_THREAD_ID);
```
There is a new thread naming scheme. See #8216 Can you set the name accordingly?

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/resources/ESH-INF/config/config.xml:
```diff
@@ -0,0 +1,47 @@
+			<context>password</context>
+			<label>Password</label>
+			<description>Password for TouchWand Wanderfull Hub </description>
+			<required>true</required>
```
The `required` tag is deprecated. Please use the `required="true"` attribute in the `parameter` tag. Same for below.

roieg commented:
@fwolter 
I fixed all the new comments 
regarding https://github.com/openhab/openhab-addons/pull/6310#discussion_r485944641
      try {
            url = new URL(command);
        } catch (MalformedURLException e) {
            logger.warn("Error building URL {} : {}", command, e.getMessage());

if I am not implementing try catch MalformedURLException   i get an 
unhanded exception error. 
Any suggestion on this one ?

fwolter commented:
> if I am not implementing try catch MalformedURLException i get an
> unhanded exception error.
> Any suggestion on this one ?

You could declare `sendCommand()` with `throws MalformedURLException`.

fwolter commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandBaseUnitHandler.java:
```diff
@@ -0,0 +1,153 @@
+
+        unitId = getThing().getUID().getId(); // TouchWand unit id
+
+        bridgeHandler.registerUpdateListener(this);
```
As `bridgeHandler` is Nullable now, you need to check if it is not null before accessing it. You can store it to a local variable and do the null check on that. Same for the others. Please check all compiler warnings.

roieg replied:
Done for all warnings

fwolter reviewed (approved):
LGTM

roieg commented on bundles/org.openhab.binding.touchwand/README.md:
```diff
@@ -0,0 +1,80 @@
+
+**Touchwand Wanderfull™** Hub Controller need to be added manually by IP address. The controller requires **username** and **password**  
+
+| Parameter | Description| Units| required |
```
done

roieg commented on bundles/org.openhab.binding.touchwand/src/main/resources/ESH-INF/config/config.xml:
```diff
@@ -0,0 +1,45 @@
+			<label>Network Address</label>
+			<description>Network address of this TouchWand Wanderfull Hub </description>
+		</parameter>
+		<parameter name="port" type="integer" min="80" max="65535">
```
done

roieg commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandHandlerFactory.java:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    @Reference
+    protected void setHttpClientFactory(HttpClientFactory httpClientFactory) {
```
@Hilbrand  , I am not sure I totally understand the requested change 
is there an example you familiar with I can use ?

roieg commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandControllerDiscoveryService.java:
```diff
@@ -0,0 +1,142 @@
+    @Override
+    protected void startScan() {
+        DatagramSocket localListenSocket = listenSocket;
+        runReceiveThread(localListenSocket);
```
@Hilbrand , my main concern was not to block the entire process while the discovery waiting for a UDP broadcast that may not arrive

roieg commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/dto/TouchWandUnitData.java:
```diff
@@ -0,0 +1,237 @@
+    private int epId;
+    @SerializedName("icon")
+    @Expose
+    private Object icon;
```
@Hilbrand ,
 when I created the Data classes,  I used automatic script to analyze a response and convert it from Json response to java class , some of the data was without any values to the script and thus the Objects

roieg replied:
Anyhow I will assign the right types and commits

roieg commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandBridgeHandler.java:
```diff
@@ -0,0 +1,161 @@
+    private boolean addSecondaryUnits;
+    private BundleContext bundleContext;
+    private @Nullable TouchWandWebSockets touchWandWebSockets;
+    private Map<String, TouchWandUnitUpdateListener> unitUpdateListeners = new ConcurrentHashMap<>();
```
@Hilbrand efficiency was actually my main concern as status updated can arrive for a specific unit and at my how there are about 100 of those . I prefer to leave the code as is , it may a little more complicated but better performance

roieg commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandBaseUnitHandler.java:
```diff
@@ -0,0 +1,161 @@
+
+        bridgeHandler = (TouchWandBridgeHandler) bridge.getHandler();
+
+        unitId = getThing().getUID().getId(); // TouchWand unit id
```
ok , this was for the auto discovery system , I stored it in the properties and got it from there

roieg commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandBridgeHandler.java:
```diff
@@ -0,0 +1,161 @@
+        Configuration config;
+
+        updateStatus(ThingStatus.UNKNOWN);
+        config = getThing().getConfiguration();
```
done

roieg commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandBridgeHandler.java:
```diff
@@ -0,0 +1,161 @@
+public class TouchWandBridgeHandler extends BaseBridgeHandler implements TouchWandUnitStatusUpdateListener {
+    public static final Set<ThingTypeUID> SUPPORTED_THING_TYPES = Collections.singleton(THING_TYPE_BRIDGE);
+    private final Logger logger = LoggerFactory.getLogger(TouchWandBridgeHandler.class);
+    private final Map<ThingUID, @Nullable ServiceRegistration<?>> discoveryServiceRegs = Collections
```
@Hilbrand just to make sure I fully understand
I implemented two discovery services . one is for all the bridges in the network , this discovery does not bound to a Thing  - this implemented in TouchWandControllerDiscoveryService
The second discovery is all the units in the already discovered and configured bride , this implements in TouchWandUnitDiscoveryService.
I assume you were referring to the second one ?

Hilbrand replied:
Yes I'm referring to `TouchWandUnitDiscoveryService`. That one is registered here in the bridge. You can remove that whole part with `discoveryServiceRegs`. See the documentation link to what is required to implement it.

roieg commented on bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandHandlerFactory.java:
```diff
@@ -0,0 +1,85 @@
+    public static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Collections.unmodifiableSet(Stream
+            .of(TouchWandBridgeHandler.SUPPORTED_THING_TYPES.stream(),
+                    TouchWandBaseUnitHandler.SUPPORTED_THING_TYPES.stream())
+            .flatMap(i -> i).collect(Collectors.toSet()));
```
done

roieg commented:
@Hilbrand @fwolter do we continue to work on this PR or to open a new one on main for OH3 ?

Hilbrand commented:
It's possible to continue work on this pr. You need to change the branch it merges against to main here on GitHub and than get your own branch in sync with main and push force. Depending on how good you git knowledge is this is doable. Otherwise just create a new pr and close this one.

roieg commented:
> It's possible to continue work on this pr. You need to change the branch it merges against to main here on GitHub and than get your own branch in sync with main and push force. Depending on how good you git knowledge is this is doable. Otherwise just create a new pr and close this one.

So lets continue here , once finish and approved I will move to main (maybe with a little help)

fwolter commented:
Please note that you need to continue with the migrated code, the openhab-bot generated above.

Hilbrand commented:
No need to use the migrated code by the bot. You can also do the migration yourself.

roieg commented:
Hi @Hilbrand @fwolter  , are there any more actions needed from my side or just wait for review ?

fwolter commented:
You need to migrate your code to OH3. You can do this either manually or use the automatically migrated code by the openhab-bot. In either way, the next step would be to file a new PR with the migrated code and close this PR. This PR won't receive any new review action by a maintainer.

Please note to file the new PR against the `main` branch.

roieg commented:
> You need to migrate your code to OH3. You can do this either manually or use the automatically migrated code by the openhab-bot. In either way, the next step would be to file a new PR with the migrated code and close this PR. This PR won't receive any new review action by a maintainer.
> 
> Please note to file the new PR against the `main` branch.

Thanks , I will do that

fwolter commented:
Closing, because binding is already merged.
